### ISSUE TITLE
Incorrect null checks

### ISSUE THREAD

**Issue by lukehutch** [dart-lang/sdk#52477]
Given the following code:

```dart
import 'package:clickdating_server/src/generated/protocol.dart';
import 'package:serverpod/serverpod.dart';
import 'package:serverpod_auth_server/module.dart';

class AdminEndpoint extends Endpoint {
  @override
  get requireLogin => true;

  @override
  get requiredScopes => {Scope.admin};

  Future<void> banUserById(Session session, int userId) async {
    await session.auth.signOutUser(userId: userId);
    await Users.invalidateCacheForUser(session, userId);
    final user = await Users.findUserByUserId(session, userId);
    assert(user != null);
    user!.blocked = true;     // (1) `!` is required here
    await session.db.update(user);
    final profile = await Profile.findSingleRow(session,
        where: (profileTable) => profileTable.userId.equals(userId));
    assert(profile != null);
    profile!.accountStatus = 'B';  // (2) `!` is required here
    profile!.accountStatusChangedDate = DateTime.now();   // (3) `!` is not needed here
    await Profile.update(session, profile!);    // (4) `!` is not needed here
  }
}
```

See the commented lines.

For (1) and (2), really the linter or compiler should do a better job of realizing the assertion guarantees the local is not null after the assertion.

But even if that's not possible (because it is Dart's policy to not rely on assertions for null analysis, e.g. if assertion checks can be disabled), then still (3) and (4) should not be reporting `The '!' will have no effect because the receiver can't be null. Try removing the '!' operator.`

(2) is inconsistent with (3) / (4).

* Dart 3.0.1, Linux

**Comment by water-mizuu** [comment]
The thing is, asserts can be disabled. So basically asserts aren't guaranteed to run, therefore it can lead to unsafe behavior.

**Comment by water-mizuu** [comment]
> (2) is inconsistent with (3)/(4).

Basically when you use a null-assert operator `!`, it promotes the variable to non-nullable, because there are two cases.

First case, the null-assert operator throws an Exception. Basically, the code after it doesn't run because it *is* null.
Second case, the null-assert operator succeeds *guaranteeing* to the compiler that the value is not null. So it doesn't force you to null assert again after (2). See here:

```dart
import 'package:clickdating_server/src/generated/protocol.dart';
import 'package:serverpod/serverpod.dart';
import 'package:serverpod_auth_server/module.dart';

class AdminEndpoint extends Endpoint {
  @override
  get requireLogin => true;

  @override
  get requiredScopes => {Scope.admin};

  Future<void> banUserById(Session session, int userId) async {
    // ...
    final profile = await Profile.findSingleRow(session,
        where: (profileTable) => profileTable.userId.equals(userId));
    assert(profile != null);
    // If [profile] is [null], the program will throw an exception, ignoring the code below.
    // else it's not [null], and we continue running, guaranteeing that it's not null anymore,
    //   so we don't need the assert in the following lines after this.
    profile!.accountStatus = 'B';  // (2) `!` is required here
    //  (You *can* remove it now, because it got promoted by the compiler to not null).
    profile!.accountStatusChangedDate = DateTime.now();   // (3) `!` is not needed here
    await Profile.update(session, profile!);    // (4) `!` is not needed here
  }
}
```

Also, asserts aren't guaranteed to run in production, so it's unsafe to use them as null assertions. (Which you noted.)

**Comment by lukehutch** [comment]
> `    //  (You *can* remove it now, because it got promoted by the compiler to not null).`

No, I can't remove the `!` at (2), it is required. Presumably because asserts can be disabled.

However I solved the mystery as to why the `!` is not needed at (3) or (4): those two instances are guarded by (2). I thought it was inconsistent behavior, but if I remove the `!` from (3) and (4), there is no error, but if I then remove the `!` from (2), there is an error on all 3 lines, because there is now no guard.

I hadn't taken the guarding behavior into consideration, which is why I found this fishy. But now it makes perfect sense. Thanks, and sorry to take your time on this!
